lib/frp/src/io/keyboard.rs

```diff
@@ -0,0 +1,264 @@
+//! Keboard FRP bindings.
+use crate::prelude::*;
+
+use crate::*;
+use rust_dense_bitset::BitSet;
+use rust_dense_bitset::DenseBitSetExtended;
+use std::collections::hash_map::Entry;
+use crate::core::fmt::{Formatter, Error};
+
+
+// ===========
+// === Key ===
+// ===========
+
+/// A type of key.
+pub use keyboard_types::Key;
+
+///// A keyboard key representation. This newtype extends the `Key` structure taken from
+///// `keyboard_types` crate, e.g. adding possibility to create Key from String,
+///// and implements `Default`.
+//#[derive(Clone,Debug,Eq,FromStr,PartialEq,Shrinkwrap)]
+//pub struct Key(pub KeyType);
+
+
+// ===============
+// === KeyMask ===
+// ===============
+
+/// The assumed maximum key code, used as size of KeyMask bitset.
+const MAX_KEY_CODE : usize = 255;
+
+/// The key bitmask - each bit represents one key. Used for matching key combinations.
+#[derive(BitXor,Clone,Debug,Eq,Hash,PartialEq,Shrinkwrap)]
+#[shrinkwrap(mutable)]
+pub struct KeyMask(pub DenseBitSetExtended);
+
+impl KeyMask {
+    /// Check if key bit is on.
+    pub fn has_key(&self, key:&Key) -> bool {
+        let KeyMask(bit_set) = self;
+        bit_set.get_bit(key.legacy_keycode() as usize)
+    }
+}
+
+impl Default for KeyMask {
+    fn default() -> Self {
+        let mut bitset = DenseBitSetExtended::with_capacity(MAX_KEY_CODE + 1);
+        // This is the only way to set bitset length.
+        bitset.set_bit(MAX_KEY_CODE,true);
+        bitset.set_bit(MAX_KEY_CODE,false);
+        Self(bitset)
+    }
+}
+
+impl<'a> FromIterator<&'a Key> for KeyMask {
+    fn from_iter<T: IntoIterator<Item=&'a Key>>(iter:T) -> Self {
+        let mut key_mask = KeyMask::default();
+        for key in iter {
+            let bit = key.legacy_keycode() as usize;
+            key_mask.set_bit(bit,true);
+        }
+        key_mask
+    }
+}
+
+impl From<&[Key]> for KeyMask {
+    fn from(keys: &[Key]) -> Self {
+        <KeyMask as FromIterator<&Key>>::from_iter(keys)
+    }
+}
+
+
+// ================
+// === KeyState ===
+// ================
+
+/// A helper structure used for describing specific key state.
+#[derive(Clone,Debug,Default)]
+struct KeyState {
+    key     : Key,
```

Comment:
How on earth the `Key` is not Copy? I was looking into this library and I dont get it. The design imo is stupid. Can we do anything with it ? :/

Reply:
Probably it's not copy, because one variant contains String.